sys/net/network_layer/ng_ipv6/ext/ng_ipv6_ext.c

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright (C) 2015 Martine Lenders <mlenders@inf.fu-berlin.de>
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+/**
+ * @{
+ *
+ * @file
+ */
+
+#include <errno.h>
+
+#include "utlist.h"
+#include "net/ng_pktbuf.h"
+#include "net/ng_ipv6.h"
+
+#include "net/ng_ipv6/ext.h"
+
+bool ng_ipv6_ext_demux(kernel_pid_t iface, ng_pktsnip_t *pkt,
+                       uint8_t nh)
+{
+    ng_pktsnip_t *ext_snip;
+    ng_ipv6_ext_t *ext;
+    unsigned int offset = 0;
+
+    ext = ((ng_ipv6_ext_t *)(((uint8_t *)pkt->data) + sizeof(ng_ipv6_hdr_t)));
+
+    bool c = true;
+
+    while (c) {
+        switch (nh) {
+            case NG_PROTNUM_IPV6_EXT_HOPOPT:
+            case NG_PROTNUM_IPV6_EXT_DST:
+            case NG_PROTNUM_IPV6_EXT_RH:
+            case NG_PROTNUM_IPV6_EXT_FRAG:
+            case NG_PROTNUM_IPV6_EXT_AH:
+            case NG_PROTNUM_IPV6_EXT_ESP:
+            case NG_PROTNUM_IPV6_EXT_MOB:
+                /* TODO: add handling of types */
+                nh = ext->nh;
+                offset += ((ext->len * NG_IPV6_EXT_LEN_UNIT) + NG_IPV6_EXT_LEN_UNIT);
+                ext = ng_ipv6_ext_get_next((ng_ipv6_ext_t *)ext);
+                break;
+
+            default:
+                c = false;
```

Comment:
Okay I understand now that the `default case` in this switch is for unrecognized exntension headers **and** the first appearance of a valid next header other than an extension header. This was a little bit strange and not really intuitive for me. I guess the packet is discarded in a later stage when it tries to interpret the unrecognized next header.